File: python/rest_api/rest_api/routes_posts.py
from fastapi import APIRouter, HTTPException, Query

from .models import PaginatedResponse, Post, PostCreate, PostUpdate
from .store import store

router = APIRouter(prefix="/posts", tags=["posts"])


@router.get("", response_model=PaginatedResponse[Post])
def list_posts(
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    author_id: int | None = None,
    published: bool = False,
) -> PaginatedResponse[Post]:
    return store.list_posts(page, per_page, author_id, published)


@router.post("", response_model=Post, status_code=201)
def create_post(body: PostCreate, author_id: int = Query(...)) -> Post:
    assert store.get_user(author_id) is not None, f"Author {author_id} not found"
    return store.create_post(author_id, body)


@router.get("/{post_id}", response_model=Post)
def get_post(post_id: int) -> Post:
    post = store.get_post(post_id)
    if post is None:
        raise HTTPException(404, "Post not found")
    return post


@router.patch("/{post_id}", response_model=Post)
def update_post(post_id: int, body: PostUpdate) -> Post:
    post = store.update_post(post_id, body)
    if post is None:
        raise HTTPException(404, "Post not found")
    return post


@router.delete("/{post_id}", status_code=204)
def delete_post(post_id: int) -> None:
    if not store.delete_post(post_id):
        raise HTTPException(404, "Post not found")

File: elixir/rest_api/lib/rest_api/routes_posts.ex
defmodule RestApi.RoutesPosts do
  use Plug.Router
  alias RestApi.{Models.Post, Store, Router}

  plug(:match)
  plug(:dispatch)

  get "/" do
    conn = fetch_query_params(conn)
    page = parse_int(conn.query_params["page"], 1)
    per_page = parse_int(conn.query_params["per_page"], 20)
    author_id = parse_optional_int(conn.query_params["author_id"])
    published = conn.query_params["published"] == "true"
    result = Store.list_posts(page, per_page, author_id, published)
    result = %{result | items: Enum.map(result.items, &Post.to_json/1)}
    Router.send_json(conn, 200, result)
  end

  post "/" do
    conn = fetch_query_params(conn)
    author_id = String.to_integer(conn.query_params["author_id"])
    true = Store.get_user(author_id) != nil
    post = Store.create_post(author_id, conn.body_params)
    Router.send_json(conn, 201, Post.to_json(post))
  end

  get "/:id" do
    case Store.get_post(String.to_integer(id)) do
      nil -> Router.send_json(conn, 404, %{detail: "Post not found"})
      post -> Router.send_json(conn, 200, Post.to_json(post))
    end
  end

  patch "/:id" do
    case Store.update_post(String.to_integer(id), conn.body_params) do
      nil -> Router.send_json(conn, 404, %{detail: "Post not found"})
      post -> Router.send_json(conn, 200, Post.to_json(post))
    end
  end

  delete "/:id" do
    case Store.delete_post(String.to_integer(id)) do
      false -> Router.send_json(conn, 404, %{detail: "Post not found"})
      true -> send_resp(conn, 204, "")
    end
  end

  defp parse_int(nil, default), do: default
  defp parse_int(val, _default), do: String.to_integer(val)

  defp parse_optional_int(nil), do: nil
  defp parse_optional_int(val), do: String.to_integer(val)
end
